### PR TITLE
Create dataproc serverless spark batches operator

### PR DESCRIPTION
**Pull request by MaksYermak**
Create operator for working with Batches for Google Dataproc. Includes operators, hooks, example dags, tests and docs.

Co-authored-by: Wojciech Januszek januszek@google.com
Co-authored-by: Lukasz Wyszomirski wyszomirski@google.com
Co-authored-by: Maksim Yermakou maksimy@google.com

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/providers/google/cloud/operators/dataproc.py
Comment (josh-fell) [inline airflow/providers/google/cloud/operators/dataproc.py:2202]: Missing `gcp_conn_id` and `impersonation_chain` from the docstring.

File: airflow/providers/google/cloud/example_dags/example_dataproc.py
Comment (josh-fell) [inline airflow/providers/google/cloud/example_dags/example_dataproc.py]:
```suggestion
    "example_gcp_batch_dataproc",
    schedule_interval='@once',
    start_date=datetime(2021, 1, 1),
    catchup=False,
```

There is an almost-finished effort to transition away from using `start_date=days_ago(n)` in example DAGs to using a static value as best practice. New example DAGs should follow the static start_date approach. The value used doesn't matter as long as it's static.

Also adding `catchup=False` has been recently discussed for all example DAGs to help with any accidental DAG run explosions if new users mutate the `schedule_interval` without fully understanding `catchup=True` is the default.

File: airflow/providers/google/cloud/operators/dataproc.py
Comment (MaksYermak) [reply airflow/providers/google/cloud/operators/dataproc.py:2202]: @josh-fell I've added it

File: airflow/providers/google/cloud/example_dags/example_dataproc.py
Comment (MaksYermak) [reply airflow/providers/google/cloud/example_dags/example_dataproc.py]: @josh-fell I've changed this code to the new approach

File: airflow/providers/google/cloud/operators/dataproc.py
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/dataproc.py:2256]: Inconsistent return statements. If batch exist we should return existing batch

File: airflow/providers/google/cloud/operators/dataproc.py
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/dataproc.py:2491]: Is there a chance that this response will be bigger than default allowed XCom size? I've never thought about how operators should handle such cases. @potiuk @mik-laj WDYT?

**Comment by MaksYermak** [comment]
> * dynamically create `batch_id` values

@josh-fell answers for your questions 

1. I think it can be useful I will add it to a code.
2. No, because `DataprocCreateBatchOperator` return a `Batch` object, not a `batch_id`

**Comment by MaksYermak** [comment]
> @MaksYermak General thoughts across the new operators:
> 
> * WDYT about adding `batch_id` as a `template_field`? Would it be beneficial for users to be able to dynamically create `batch_id` values?
> * Adjacently, is `batch_id` something the `DataprocCreateBatchOperator` can push as an `XCom` such that the value can be used in downstream tasks as an `XComArgs` (i.e. the use case in the example DAG if the `.output` property is used as an input to the `DataprocGetBatchOperator` and `DataprocDeleteBatchOperator` tasks)?
> 
> Not saying it needs to/should be implemented but thought these might be good (and possibly cheap) features.

@josh-fell answers for your questions

1. I think it can be useful I will add it to a code.
2. No, because `DataprocCreateBatchOperator` return a `Batch` object, not a `batch_id`

File: airflow/providers/google/cloud/operators/dataproc.py
Comment (MaksYermak) [reply airflow/providers/google/cloud/operators/dataproc.py:2491]: @turbaszek I've fix it

File: airflow/providers/google/cloud/operators/dataproc.py
Comment (MaksYermak) [reply airflow/providers/google/cloud/operators/dataproc.py:2256]: @turbaszek I´ve fix it

**Comment by MaksYermak** [comment]
@turbaszek @josh-fell @mik-laj hi guys, could you look on this PR one more time?

**Comment by josh-fell** [comment]
@MaksYermak Should `batch_id` be added to `template_fields` in `DataprocGetBatchOperator` and `DataprocDeleteBatchOperator` as well? No strong opinion though.

LGTM 👍

**Comment by MaksYermak** [comment]
> @MaksYermak Should `batch_id` be added to `template_fields` in `DataprocGetBatchOperator` and `DataprocDeleteBatchOperator` as well? No strong opinion though.
> 
> LGTM 👍

@josh-fell make sense, I have added it to the code.

**Comment by MaksYermak** [comment]
@turbaszek @josh-fell @mik-laj hi guys, could you look and approve this PR for merge if all good?

**Comment by lwyszomi** [comment]
@turbaszek @josh-fell @mik-laj Is the a chance to merge this?

**Comment by josh-fell** [comment]
@MaksYermak @lwyszomi There is a static check that is failing. Can you address this?

FYI - I am not able to merge. A code owner will have to do that.

**Comment by MaksYermak** [comment]
> @MaksYermak @lwyszomi There is a static check that is failing. Can you address this?
> 
> FYI - I am not able to merge. A code owner will have to do that.

@josh-fell I have fixed it

File: airflow/providers/google/cloud/operators/dataproc.py
Comment (potiuk) [reply airflow/providers/google/cloud/operators/dataproc.py:2491]: I think this is not a problem - such response is unlikely to get bigger than the limit, and even if so - we are actually much more relaxed now with XCom, as custom xcom backends might handle arbitrary size of data if needed.

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Just intermittent errors. Merging.

**Comment by aoelvp94** [comment]
A question here, how can I specify the docker image to run pyspark workloads in custom containers using DataprocCreateBatchOperator?

**Comment by lwyszomi** [comment]
@aoelvp94 I'm not a expert of Dataproc serverless service, but operator is only a wrapper for the SDK and we take the same parameters as we heve there. So in the `Batch` definision we have `runtimeConfig` where you can specify the image. I hope this will help you.

Refs:
Batch -> https://cloud.google.com/dataproc-serverless/docs/reference/rest/v1/projects.locations.batches#Batch
runtimeConfig -> https://cloud.google.com/dataproc-serverless/docs/reference/rest/v1/projects.locations.batches#RuntimeConfig

**Comment by aoelvp94** [comment]
Yeah @lwyszomi I am trying with something like that. I am trying that now so if I have news I will leave the code snippet for future references

**Comment by aoelvp94** [comment]
@lwyszomi No success when I tried to set jinja params in the `Batch` object, did you know why?

**Comment by lwyszomi** [comment]
@aoelvp94 which version of the provider you are using? The Batch object is templated field starting from 6.4.0

**Comment by aoelvp94** [comment]
> @aoelvp94 which version of the provider you are using? The Batch object is templated field starting from 6.4.0

I am using `composer-2.0.5-airflow-2.2.3` so I have 6.4.0

**Comment by aoelvp94** [comment]
Another question, why the operator don't generate a hash in `batch_id` field as happens in `DataprocSubmitPySparkJobOperator` (I am replacing this implementation)? I have to delete the batch to run again the same batch (I prefer having the historical runs)

**Comment by lwyszomi** [comment]
> > @aoelvp94 which version of the provider you are using? The Batch object is templated field starting from 6.4.0
> 
> I am using `composer-2.0.5-airflow-2.2.3` so I have 6.4.0

I will need more investigation why jijna paramams doesn't work, I checked and you have right that for Composer 2.0.5 using 6.4.0.

> Another question, why the operator don't generate a hash in `batch_id` field as happens in `DataprocSubmitPySparkJobOperator` (I am replacing this implementation)? I have to delete the batch to run again the same batch (I prefer having the historical runs)

We created operators based on the exisiting SDK, we didn't add any extra logic to add hash to the `batch_id`.

**Comment by lwyszomi** [comment]
@aoelvp94 for which property you want to use jinja template?

**Comment by aoelvp94** [comment]
`batch` when I use `Batch` / `PySparkBatch` object

**Comment by lwyszomi** [comment]
@aoelvp94 butch is tepletized so it should work, but which property inside `Batch/PySparkBatch` you try to templetize

**Comment by lwyszomi** [comment]
can you share Batch config?

**Comment by aoelvp94** [comment]
image
I removed the main python file on purpose.

```python
def get_week_ago(dt):
    bd = dt + relativedelta(days=-7)
    return bd.strftime('%Y-%m-%d')

dag = DAG(
        dag_id=f"{table_name}_table_v4",
        ...
        user_defined_macros={'week_ago': get_week_ago}
    )


create_batch = DataprocCreateBatchOperator(
        task_id="create_batch",
        project_id=project_id,
        region=f"{{{{ var.value.region_{environment} }}}}",
        batch=Batch(pyspark_batch=PySparkBatch(
        main_python_file_uri=main_python_file_uri,
        python_file_uris=files,
        args=[
            "--start-date",
            "{{ week_ago(execution_date) }}", # it doesn't work
            "--end-date",
            "{{ ds }}", # it doesn't work
            "--table",
            table_name,
            "--source-bucket",
            source_bucket,
            "--source-prefix",
            source_prefix,
            "--output-path",
            f"{output_path}/{table_name}/",
        ],
    ), runtime_config=RuntimeConfig(container_image="gcr.io/company/image_name:1.0.1")),
        batch_id=f"pyspark-table-{table_name.replace('_','-')}-{{{{ ds_nodash }}}}",
        retry=DEFAULT,
        timeout=500,
    )
```

**Comment by lwyszomi** [comment]
@aoelvp94 thanks, we will check why this not work, I will back to you with update when I will have any information

**Comment by marekw-openx** [comment]
@aoelvp94, have you tried passing `batch` config in the form of python dict? in my case I'm using just a python dict and it gets templetized, like:
```
batch = {
  "spark_batch": {
    "spark_job": {
      "args": ["-d", "{{ ts }}"]
    }
  }
  (...)
}
```

also, please double check the apache-airflow-providers-google version (`batch` wasn't templated in the 6.2.0)

**Comment by MaksYermak** [comment]
@aoelvp94 I have checked your configuration for correct work you should use dictionary instead of Batch() object. It is because in some reasons Airflow can't template object's property.

`{
            "pyspark_batch": {
                "main_python_file_uri": main_python_file_uri,
                "python_file_uris": files,
                "args": [
                    "--start-date={{ get_week_ago(data_interval_start) }}",
                    "--end-date={{ ds }}",
                    (. . .)
                ],
            },
           ( . . .)
}`
One more thing in the last Jinja version 'execution_date' from the template is deprecated. Please use 'data_interval_start' or 'logical_date' instead.

**Pull request merged**
